x-pack/plugins/security_solution/server/lib/telemetry/endpoint.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ISavedObjectsRepository } from 'src/core/server';
+import { AgentMetadata } from '../../../../ingest_manager/common/types/models/agent';
+import { getFleetSavedObjectsMetadata, getFleetEventsSavedObjects } from './fleet_saved_objects';
+
+export interface AgentOSMetadataTelemetry {
+  full_name: string;
+  platform: string;
+  version: string;
+  count: number;
+}
+
+export interface PoliciesTelemetry {
+  malware: {
+    success: number;
+    warning: number;
+    failure: number;
+  };
+}
+
+export interface AgentMetadataTelemetry {
+  total_installed: number;
+  active_within_last_24_hours: number;
+  os: AgentOSMetadataTelemetry[];
+  policies?: PoliciesTelemetry; // TODO: make required when able to enable policy information
+}
+
+export interface AgentLocalMetadata extends AgentMetadata {
+  elastic: {
+    agent: {
+      id: string;
+    };
+  };
+  host: {
+    id: string;
+  };
+  os: {
+    name: string;
+    platform: string;
+    version: string;
+    full: string;
+  };
+}
+
+export const getDefaultEndpointTelemetry = (): AgentMetadataTelemetry => ({
+  total_installed: 0,
+  active_within_last_24_hours: 0,
+  os: [],
+});
+
+export const getEndpointTelemetryFromFleet = async (
+  savedObjectsClient: ISavedObjectsRepository
+) => {
+  // Retrieve every agent that references the endpoint as an installed package. It will not be listed if it was never installed
+  const { saved_objects: endpointAgents } = await getFleetSavedObjectsMetadata(savedObjectsClient);
+  const endpointTelemetry = getDefaultEndpointTelemetry();
+  // If there are no installed endpoints return the default telemetry object
+  if (!endpointAgents || endpointAgents.length < 1) return endpointTelemetry;
+
+  const uniqueHostIds: Set<string> = new Set();
+  const uniqueAgentIds: Set<string> = new Set();
+
+  // All agents who have the endpoint registered as an installed package.
+  endpointTelemetry.total_installed = endpointAgents.length;
+
+  const aDayAgo = new Date();
+  aDayAgo.setDate(aDayAgo.getDate() - 1);
+
+  const osTracker: Record<string, AgentOSMetadataTelemetry> = {};
+
+  const endpointMetadataTelemetry = endpointAgents.reduce(
+    (metadataTelemetry, { attributes: metadataAttributes }) => {
+      const { last_checkin: lastCheckin, local_metadata: localMetadata } = metadataAttributes;
+      const { host, os, elastic } = localMetadata as AgentLocalMetadata;
+
+      if (lastCheckin && new Date(lastCheckin) > aDayAgo) {
```

Comment:
:grey_question: This is a big `reduce` and it seems like maybe it could be easier to read if we did some of this stuff in front of it as a `filter`. This is a :grey_question: more on the "ignore for now" side, just a comment for later.

Reply:
Will break apart in next commit